internal/trie/node/header.go

```diff
@@ -132,24 +151,29 @@ var ErrVariantUnknown = errors.New("node variant is unknown")
 // the decodeHeaderByte function below.
 // For 7 variants, the performance is improved by ~20%.
 var variantsOrderedByBitMask = [...]variant{
-	leafVariant,            // mask 1100_0000
-	branchVariant,          // mask 1100_0000
-	branchWithValueVariant, // mask 1100_0000
+	leafVariant,                  // mask 1100_0000
+	branchVariant,                // mask 1100_0000
+	branchWithValueVariant,       // mask 1100_0000
+	leafWithHashedValueVariant,   // mask 1110_0000
+	branchWithHashedValueVariant, // mask 1111_0000
+	emptyVariant,                 // mask 1111_1111
+	compactEncodingVariant,       // mask 1111_1111
 }
 
-func decodeHeaderByte(header byte) (variantBits,
-	partialKeyLengthHeader, partialKeyLengthHeaderMask byte, err error) {
+func decodeHeaderByte(header byte) (nodeVariant variant,
+	partialKeyLengthHeader byte, err error) {
+	var partialKeyLengthHeaderMask byte
 	for i := len(variantsOrderedByBitMask) - 1; i >= 0; i-- {
-		variantBits = header & variantsOrderedByBitMask[i].mask
-		if variantBits != variantsOrderedByBitMask[i].bits {
+		nodeVariant = variantsOrderedByBitMask[i]
+		variantBits := header & nodeVariant.mask
+		if variantBits != nodeVariant.bits {
 			continue
 		}
 
-		partialKeyLengthHeaderMask = ^variantsOrderedByBitMask[i].mask
+		partialKeyLengthHeaderMask = nodeVariant.partialKeyLengthHeaderMask()
 		partialKeyLengthHeader = header & partialKeyLengthHeaderMask
-		return variantBits, partialKeyLengthHeader,
-			partialKeyLengthHeaderMask, nil
+		return nodeVariant, partialKeyLengthHeader, nil
 	}
 
-	return 0, 0, 0, fmt.Errorf("%w: for header byte %08b", ErrVariantUnknown, header)
+	return variant{}, 0, fmt.Errorf("%w: for header byte %08b", ErrVariantUnknown, header)
```

Comment:
maybe define an `invalidVariant` so we don't just keep passing around zero value variants?